[red-knot] Fix merged type after if-else without explicit else branch (#14621)
astral-sh/ruff: crates/red_knot_python_semantic/resources/mdtest/narrow/post_if_statement.md

# Consolidating narrowed types after if statement

> ## After if-else statements, narrowing has no-effect if variable is not mutated

Reviewer: ```suggestion
## After if-else statements, narrowing has no effect if the variable is not mutated in any branch
```

```py
def optional_int() -> int | None: ...

x = optional_int()

if x is None:
    pass
else:
    pass

reveal_type(x)  # revealed: int | None
```

> ## After if-else statements, narrowing has an effect if variable is mutated

Reviewer: ```suggestion
## Narrowing can have a persistent effect if the variable is mutated in one branch
```

```py
def optional_int() -> int | None: ...

x = optional_int()

> y = optional_int()

Reviewer: It looks like `y` is not used in this test?
```suggestion
```

if x is None:
    x = 10
else:
    pass

reveal_type(x)  # revealed: int
```

> ## if statement without explicit else branch, act similar to empty else branch

Reviewer: ```suggestion
## An if statement without an explicit `else` branch is equivalent to one with a no-op `else` branch
```

```py
def optional_int() -> int | None: ...

x = optional_int()
y = optional_int()

if x is None:
    x = 0

if y is None:
    pass

reveal_type(x)  # revealed: int
reveal_type(y)  # revealed: int | None
```

> ## if-elif without explicit else branch, act similar to empty else branch

Reviewer: ```suggestion
## An if-elif without an explicit else branch is equivalent to one with an empty else branch
```
